pkg/router/controller/status.go

```diff
@@ -146,30 +179,87 @@ func performIngressConditionUpdate(action string, lease writerlease.Lease, track
+			}
+			return writerlease.None, false
+		}
+
+		// If the tracker determines that another process is attempting to update the ingress to an inconsistent
+		// value, skip updating altogether and rely on the next resync to resolve conflicts. This prevents routers
+		// with different configurations from endlessly updating the route status.
+		if tracker.IsChangeContended(key, now, latest) {
+			log.V(4).Info("skipped update due to another process altering the route with a different ingress status value", "action", action, "key", key, "original", original)
 			return writerlease.Release, false
-		case errors.IsConflict(err):
-			// just follow the normal process, and retry when we receive the update notification due to
-			// the other entity updating the route.
-			log.V(4).Info("updating route status failed due to write conflict", "action", action, "namespace", route.Namespace, "name", route.Name)
-			return writerlease.Release, true
-		default:
-			utilruntime.HandleError(fmt.Errorf("Unable to write router status for %s/%s: %v", route.Namespace, route.Name, err))
-			return writerlease.Release, true
 		}
+
+		return handleRouteStatusUpdate(key, action, oc, route, latest, tracker)
 	})
 }
 
+// handleRouteStatusUpdate manages the update of route status in conjunction with a writerlease and a tracker. It
+// attempts to update the route status and, depending on the outcome, clears the tracker if necessary. It returns the
+// writerlease's WorkResult and a boolean flag indicating whether the writerlease should retry.
+func handleRouteStatusUpdate(key, action string, oc client.RoutesGetter, route *routev1.Route, latest *routev1.RouteIngress, tracker ContentionTracker) (workResult writerlease.WorkResult, retry bool) {
+	switch _, err := oc.Routes(route.Namespace).UpdateStatus(context.TODO(), route, metav1.UpdateOptions{}); {
+	case err == nil:
+		log.V(4).Info("updated route status", "action", action, "namespace", route.Namespace, "name", route.Name)
+		tracker.Clear(key, latest)
+		return writerlease.Extend, false
+	case errors.IsNotFound(err):
+		// route was deleted
+		log.V(4).Info("route was deleted before we could update status", "action", action, "namespace", route.Namespace, "name", route.Name)
+		return writerlease.Release, false
+	case errors.IsConflict(err):
+		// just follow the normal process, and retry when we receive the update notification due to
+		// the other entity updating the route.
+		log.V(4).Info("updating route status failed due to write conflict", "action", action, "namespace", route.Namespace, "name", route.Name)
+		return writerlease.Release, true
+	default:
+		utilruntime.HandleError(fmt.Errorf("Unable to write router status for %s/%s: %v", route.Namespace, route.Name, err))
+		return writerlease.Release, true
+	}
+}
+
+// leaseKey creates a unique key for the writer lease logic given a route UID and a condition type.
+func leaseKey(uid types.UID, conditionType routev1.RouteIngressConditionType) string {
```

Comment:
Define a new type instead of using string.  This will help us make sure no one is using `route.UID` instead of `leaseKey(route.UID, condition.Type)`.

Reply:
I'm not certain I did this how you were thinking about it, but take a look. I added the new type in the writelease logic and changed the functions such as `Try(key WorkKey, fn WorkFunc)` to take `WorkKey`. Let me know if I misunderstood.